go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -63,61 +83,163 @@ func New(
+	return c
 }
 
 func (c *Controller) Run() error {
-	klog.Infof("Starting UserDefinedNetworkManager Controllers")
-	if err := controller.Start(c.nadNotifier.Controller, c.Controller); err != nil {
-		return fmt.Errorf("unable to start UserDefinedNetworkManager controller: %v", err)
+	klog.Infof("Starting user-defiend network controllers")
+	if err := controller.Start(
+		c.cudnController,
+		c.udnController,
+		c.nadNotifier.Controller,
+		c.namespaceNotifier.Controller,
+	); err != nil {
+		return fmt.Errorf("unable to start user-defiend network controller: %v", err)
 	}
 
 	return nil
 }
 
 func (c *Controller) Shutdown() {
-	controller.Stop(c.nadNotifier.Controller, c.Controller)
+	controller.Stop(
+		c.cudnController,
+		c.udnController,
+		c.nadNotifier.Controller,
+		c.namespaceNotifier.Controller,
+	)
+}
+
+// ReconcileNetAttachDef enqueue NAD requests following NAD events.
+func (c *Controller) ReconcileNetAttachDef(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return fmt.Errorf("failed to split meta namespace key %q: %v", key, err)
+	}
+	nad, err := c.nadLister.NetworkAttachmentDefinitions(namespace).Get(name)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get NetworkAttachmentDefinition %q from cache: %v", key, err)
+	}
+	ownerRef := metav1.GetControllerOf(nad)
+	if ownerRef == nil {
+		return nil
+	}
+
+	switch ownerRef.Kind {
+	case "ClusterUserDefinedNetwork":
+		owner, err := c.cudnLister.Get(ownerRef.Name)
+		if err != nil {
+			return fmt.Errorf("failed to get ClusterUserDefinedNetwork %q from cache: %v", ownerRef.Name, err)
+		}
+		ownerKey, err := cache.MetaNamespaceKeyFunc(owner)
+		if err != nil {
+			return fmt.Errorf("failed to generate meta namespace key for CUDN: %v", err)
+		}
+		c.cudnController.Reconcile(ownerKey)
+	case "UserDefinedNetwork":
+		owner, err := c.udnLister.UserDefinedNetworks(nad.Namespace).Get(ownerRef.Name)
+		if err != nil {
+			return fmt.Errorf("failed to get UserDefinedNetwork %q from cache: %v", ownerRef.Name, err)
+		}
+		ownerKey, err := cache.MetaNamespaceKeyFunc(owner)
+		if err != nil {
+			return fmt.Errorf("failed to generate meta namespace key for UDN: %v", err)
+		}
+		c.udnController.Reconcile(ownerKey)
+	default:
+		return nil
+	}
+	return nil
+}
+
+// ReconcileNamespace enqueue relevant Cluster UDN CR requests following namespace events.
+func (c *Controller) ReconcileNamespace(key string) error {
+	c.namespaceTrackerLock.RLock()
+	defer c.namespaceTrackerLock.RUnlock()
+
+	affectedNamespace, selectedNamespace := false, false
+
+	for cudnName, affectedNamespaces := range c.namespaceTracker {
+		affectedNamespace = affectedNamespaces.Has(key)
+
+		if !affectedNamespace {
+			cudn, err := c.cudnLister.Get(cudnName)
+			if err != nil {
+				return fmt.Errorf("faild to get CUDN %q from cache: %w", cudnName, err)
+			}
+			namespace, err := c.namespaceInformer.Lister().Get(key)
```

Comment:
you can do that outside the CUDN loop, then it will only be 1 call

Reply:
Done